python/pyarrow/tests/test_array.py

```diff
@@ -1090,6 +1090,57 @@ def test_invalid_tensor_construction():
         pa.Tensor()
 
 
+def test_list_array_flatten():
+    typ2 = pa.list_(
+        pa.list_(
+            pa.int64()
+        )
+    )
+    arr2 = pa.array([
+        None,
+        [
+            [1, None, 2],
+            None,
+            [3, 4]
+        ],
+        [],
+        [
+            [],
+            [5, 6],
+            None
+        ],
+        [
+            [7, 8]
+        ]
+    ])
+    assert arr2.type.equals(typ2)
+
+    typ1 = pa.list_(pa.int64())
+    arr1 = pa.array([
+        [1, None, 2],
+        None,
+        [3, 4],
+        [],
+        [5, 6],
+        None,
+        [7, 8]
+    ])
+    assert arr1.type.equals(typ1)
+
+    typ0 = pa.int64()
+    arr0 = pa.array([
+        1, None, 2,
```

Comment:
The semantics look weird to me. Why are the inner Nones kept but not the outer ones? Is this something that's desirable from the user's POV?
(note that's the question I was asking on the JIRA)

Reply:
IMO we should think None as empty lists in this scenario. The naming can be misleading though. 
How about having an `unnest` method, which removes one nesting level and a `flatten` method which unnests until a non-list value_type?

So 
- unnest: `list<list<int64>> -> list<int64>`
- flatten: `list<list<int64>> -> int64`

Reply:
I would rather have `flatten` and `flatten_recursive`, or something similar.

Reply:
Do the semantics make sense to You this way?

Reply:
Kind of. I'm not sure what kind of use case they map to, though. Since @wesm filed the original issue, it would be nice if he could clarify.

Reply:
This function should return the inner values as a zero-copy operation, that's all that was intended